File: bindings/node/src/embedded-library.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');

// The library this package ships under lib/<platform>/ for the current
// platform; null when none is bundled for it, or when lib/ is empty, as in
// a checked-out workspace.
function embeddedLibrary(platform, lib) {
  if (!platform) {
    return null;
  }
  const file = path.join(lib, platform.name, platform.file);
  return fs.existsSync(file) ? file : null;
}

module.exports = embeddedLibrary;

File: bindings/go/embedded_library.go
package memless

import (
	"io/fs"
	"path"
)

type embeddedLibrary struct {
	files   fs.FS
	file    string
	version string
}

// extract copies the bundled library into the cache and returns its path, or
// "" when there is none for this platform or the cache is unusable.
func (lib embeddedLibrary) extract(cacheDir string) string {
	if lib.files == nil || lib.file == "" || cacheDir == "" {
		return ""
	}
	data, err := fs.ReadFile(lib.files, lib.file)
	if err != nil {
		return ""
	}
	target, err := cachedCopy(cacheDir, lib.version, path.Base(lib.file), data)
	if err != nil {
		return ""
	}
	return target
}
